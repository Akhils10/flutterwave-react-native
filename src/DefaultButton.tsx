import React from 'react';
import {
  View,
  StyleSheet,
  ViewStyle,
  LayoutChangeEvent,
  TouchableHighlight,
  StyleProp,
} from "react-native";
import {colors} from './configs';

interface DefaultButtonProps {
  style?: StyleProp<ViewStyle>;
  onPress?: () => void;
  disabled?: boolean;
  children: React.ReactNode;
  isBusy?: boolean;
  onSizeChange?: (ev: {width: number; height: number}) => void;
  alignLeft?: 'alignLeft' | boolean,
}

function getBusyStyle(isBusy?: 'isBusy' | boolean) {
  if (!isBusy) {
    return {};
  }
  return styles.buttonBusy;
}

function getAlginStyle(alignLeft?: 'alignLeft' | boolean) {
  if (alignLeft) {
    return styles.buttonAlignLeft;
  }
  return {};
}

/**
 * Button base design.
 * @param param0
 */
const DefaultButton: React.FC<DefaultButtonProps> = function Button({
  style,
  onPress,
  disabled,
  children,
  isBusy,
  onSizeChange,
  alignLeft,
}) {
  const handleOnLayout = (ev: LayoutChangeEvent) => {
    const {width, height} = ev.nativeEvent.layout;
    if (onSizeChange) {
      onSizeChange({width, height});
    }
  };

  return (
    <TouchableHighlight
      underlayColor={colors.primary}
      disabled={disabled}
      onPress={onPress}
      style={[
        styles.button,
        getBusyStyle(isBusy),
        getAlginStyle(alignLeft),
        style
      ]}
      activeOpacity={1}
      onLayout={handleOnLayout}
      testID='flw-default-button'>
      <>
        {children}
        {isBusy
          ? (<View style={styles.buttonBusyOvelay} />)
          : null}
      </>
    </TouchableHighlight>
  );
};

const styles = StyleSheet.create({
  buttonBusyOvelay: {
    position: 'absolute',
    left: 0,
    top: 0,
    bottom: 0,
    right: 0,
    backgroundColor: 'rgba(255, 255, 255, 0.6)',
  },
  buttonAltBusyOvelay: {
    backgroundColor: 'rgba(255, 255, 255, 0.8)',
  },
  buttonBusy: {
    borderColor: '#FBDBA7',
  },
  buttonAltBusy: {
    borderColor: '#D0D0D5',
  },
  buttonAlignLeft: {
    justifyContent: 'flex-start',
  },
  buttonAlt: {
    backgroundColor: '#fff',
    borderColor: colors.secondary,
  },
  button: {
    paddingHorizontal: 16,
    minWidth: 100,
    height: 52,
    borderColor: colors.primary,
    borderWidth: 1,
    borderRadius: 6,
    backgroundColor: colors.primary,
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
    overflow: 'hidden',
  },
});

export default DefaultButton;
