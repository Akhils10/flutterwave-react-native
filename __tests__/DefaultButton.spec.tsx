import 'react-native';
import React from 'react';
import {Text} from 'react-native';
import renderer from 'react-test-renderer';
import DefaultButton from '../src/DefaultButton';
const BtnTestID = 'flw-default-button';

describe('<DefaultButton />', () => {
  it('renders primary button correctly', () => {
    const TestRender = renderer.create(
      <DefaultButton>
        <Text>Hello, World</Text>
      </DefaultButton>
    );
    expect(TestRender.toJSON()).toMatchSnapshot();
  });

  it('renders alt button correctly', () => {
    const TestRender = renderer.create(
      <DefaultButton alt>
        <Text>Hello, World</Text>
      </DefaultButton>
    );
    expect(TestRender.toJSON()).toMatchSnapshot();
  });

  it('renders overlay if busy', () => {
    const TestRender = renderer.create(
      <DefaultButton isBusy>
        <Text>Hello, World</Text>
      </DefaultButton>
    );
    expect(TestRender.toJSON()).toMatchSnapshot();
  });

  it('renders alt overlay if alt and busy', () => {
    const TestRender = renderer.create(
      <DefaultButton alt isBusy>
        <Text>Hello, World</Text>
      </DefaultButton>
    );
    expect(TestRender.toJSON()).toMatchSnapshot();
  });

  it('renders with align left style', () => {
    const TestRender = renderer.create(
      <DefaultButton alignLeft>
        <Text>Hello, World</Text>
      </DefaultButton>
    );
    expect(TestRender.toJSON()).toMatchSnapshot();
  });

  it('fires onSizeChange if set', () => {
    const onSizeChange = jest.fn();
    const TestRender = renderer.create(
      <DefaultButton onSizeChange={onSizeChange}>
          <Text>Hello, World</Text>
      </DefaultButton>
    );
    TestRender.root.findByProps({testID: BtnTestID}).props.onLayout({
      nativeEvent: {
        layout:{
          width: 1,
          height: 1,
        }
      }
    });
    expect(onSizeChange).toHaveBeenCalledTimes(1);
    expect(onSizeChange).toHaveBeenCalledWith({width: 1, height: 1});
  });

  it('fires onPress if set', () => {
    const onPress = jest.fn();
    const TestRender = renderer.create(
      <DefaultButton onPress={onPress}>
        <Text>Hello, World</Text>
      </DefaultButton>
    );

    TestRender.root.findByProps({testID: BtnTestID}).props.onPress();

    expect(TestRender.root.findByProps({testID: BtnTestID}).props.onPress).toBeDefined();
    expect(onPress).toHaveBeenCalledTimes(1);
  });
});
