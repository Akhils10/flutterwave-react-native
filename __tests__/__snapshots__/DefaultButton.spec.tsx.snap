// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DefaultButton /> renders overlay if busy 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#f5a623",
        "borderColor": "#f5a623",
        "borderRadius": 6,
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 52,
        "justifyContent": "center",
        "minWidth": 100,
        "overflow": "hidden",
        "paddingHorizontal": 16,
      },
      Object {
        "borderColor": "#FBDBA7",
      },
      Object {},
      undefined,
    ]
  }
  testID="flw-default-button"
>
  <Text>
    Hello, World
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "rgba(255, 255, 255, 0.6)",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</View>
`;

exports[`<DefaultButton /> renders primary button correctly 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#f5a623",
        "borderColor": "#f5a623",
        "borderRadius": 6,
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 52,
        "justifyContent": "center",
        "minWidth": 100,
        "overflow": "hidden",
        "paddingHorizontal": 16,
      },
      Object {},
      Object {},
      undefined,
    ]
  }
  testID="flw-default-button"
>
  <Text>
    Hello, World
  </Text>
</View>
`;

exports[`<DefaultButton /> renders with align left style 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onLayout={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#f5a623",
        "borderColor": "#f5a623",
        "borderRadius": 6,
        "borderWidth": 1,
        "flexDirection": "row",
        "height": 52,
        "justifyContent": "center",
        "minWidth": 100,
        "overflow": "hidden",
        "paddingHorizontal": 16,
      },
      Object {},
      Object {
        "justifyContent": "flex-start",
      },
      undefined,
    ]
  }
  testID="flw-default-button"
>
  <Text>
    Hello, World
  </Text>
</View>
`;
